### DISCUSSION TITLE
Finetuning VITS-VCTK to a new voice -- evalspeech_comparison

### DISCUSSION THREAD

**Discussion by skol101** [coqui-ai/TTS#854] [General Q&A] [answered]
Running for 22k steps so far -- generated sentences are getting better and better. But here's something I've observed is that  generated speech is almost a flip of the groundtruth speech. How to interpret this? 

<img width="1464" alt="Screenshot 2021-10-01 at 18 11 13" src="https://user-images.githubusercontent.com/53978091/135668859-e8fd5336-63f9-4dff-a339-9e6b82f5c0b8.png">
<img width="704" alt="Screenshot 2021-10-01 at 21 22 52" src="https://user-images.githubusercontent.com/53978091/135668876-21a2f054-6c35-49b0-85e4-a93d980218b1.png">

**Comment by erogol** [comment, marked as answer]
I think it is just a coincidence. If it is persistently the same even after a couple of epochs, let's discuss it again.

**Reply by skol101** [reply]
@erogol cheers for replying.

A question: why the fine-tuned models are so much larger than original pre-trained model, it's 300 MBytes for pre-trained model vs 997 MBytes for every checkpoint in fine tuning. 

<img width="732" alt="Screenshot 2021-10-05 at 8 50 01" src="https://user-images.githubusercontent.com/53978091/135967761-e4a3f870-f4cc-4073-a3b7-53b5704b9cfe.png">

**Reply by erogol** [reply]
No idea. srry

**Reply by lexkoro** [reply]
Guess the pre-trained model you are using is without optimizer and or discriminator? 300mb  sounds like it's only the generator.

**Reply by skol101** [reply]
@lexkoro I've downloaded the vits-vctk pre-trained model.  I assumed since it's End-to-end model, it doesn't need anything else.

Here's the directory listing.
<img width="626" alt="Screenshot 2021-10-05 at 17 00 21" src="https://user-images.githubusercontent.com/53978091/136038309-155a5111-2e20-4088-a34e-1aa28d539eb9.png">

**Reply by erogol** [reply]
yeah right point. The model is with only the generator.

**Reply by skol101** [reply]
@erogol so do I have to train the vits-vctk model from the scratch, and only thereafter I can finetune to my particular speakers?

**Reply by erogol** [reply]
You can finetune the generator starting with a new discriminator.

**Reply by skol101** [reply]
@erogol , got it, though I was merely following https://tts.readthedocs.io/en/latest/finetuning.html to train on a new speaker with the command

`CUDA_VISIBLE_DEVICES="0" python recipes/skol101/train_vits_vctk.py \
    --restore_path  /home/ubuntu/.local/share/tts/tts_models--en--vctk--vits`

trains_vits_vctk is as a copy of train_vits.py from ljspeech with minor adaptations per specific dataset. Since the VITS model already contains HiFi discriminator I thought I would only need to restore from pretrained model. Looks like I only need to teach the pretrained model to discriminate on the new speaker.

Also, there's a minor bug when --restore_path points to directory and not a model file.
`IsADirectoryError: [Errno 21] Is a directory: '/home/ubuntu/.local/share/tts/tts_models--en--vctk--vits'`

**Reply by erogol** [reply]
Right the path is pointing wrong. Thx ;) 

It should be just fine if you restore the released VITS model. 🐸TTS automatically passes all the matching layers to the new model and starts the rest from zero. 

But I've not tried VITS for fine-tuning for myself. Pls keep me in the loop and share the progress 👍

**Reply by skol101** [reply]
@erogol it's still not clear how discriminator is trained in the case of VITS model:
- When I restore_path do I actually train discriminator on a new voice? 
- How do I add speaker ID to the list of speakers IDs to be used later in the inference?

**Reply by erogol** [reply]
> When I restore_path do I actually train discriminator on a new voice?

What do you mean?

> How do I add speaker ID to the list of speakers IDs to be used later in the inference?

You cannot use a speaker that was not in the training set.

**Reply by skol101** [reply]
> > When I restore_path do I actually train discriminator on a new voice?
> 
> What do you mean?

I'm following this instruction https://tts.readthedocs.io/en/latest/finetuning.html to add new voices to pre-trained model. 
I'm using trains_vits.py with custom dataset formated per VCTK. 
I've downloaded pretrained 1M step vits_vctk model. 

> 
> > How do I add speaker ID to the list of speakers IDs to be used later in the inference?
> 
> You cannot use a speaker that was not in the training set.

Ok, but considering what I wrote here above, how do I fine tune an existing pre-trained model to new voices that can be used at inference?

**Reply by erogol** [reply]
If your dataset is formatted as VCTK. Use VCTK as your dataset selection in your config and use --restore_path. It would start with a new speaker embedding layer but restore the rest of the model.

**Reply by skol101** [reply]
@erogol that's exactly what I'm doing. The only question is how to I train on multiple new speakers and then how do I use the resulting fine tuned model at interference. Do I update speaker_idx.json with new speakers ids, I guess?

**Reply by erogol** [reply]
speaker file is created automatically in the training folder you don't need to edit it.

**Reply by skol101** [reply]
@erogol as you can see from the screenshot, there's no speaker file in the training folder created after I start training from restore_path
<img width="796" alt="Screenshot 2021-10-13 at 17 52 41" src="https://user-images.githubusercontent.com/53978091/137158212-46072cc2-6a17-4324-994c-11be3e44f84d.png">

**Reply by erogol** [reply]
If there is no speaker file it is not in multispeaker mode. Make sure you set the config right.

**Reply by erogol** [reply]
Ill try to write a tutorial on multispeaker training hopefully tomorrow. Maybe that would help.

**Reply by skol101** [reply]
That's my config.json file. Will be waiting for your tutorial.

`import os

from TTS.config.shared_configs import BaseAudioConfig
from TTS.trainer import Trainer, TrainingArgs, init_training
from TTS.tts.configs import BaseDatasetConfig, VitsConfig

output_path = os.path.dirname(os.path.abspath(__file__))
dataset_config = BaseDatasetConfig(
    name="vctk", path=os.path.join(output_path, "../../../datasets/en-us")
)
audio_config = BaseAudioConfig(
    sample_rate=22050,
    win_length=1024,
    hop_length=256,
    num_mels=80,
    preemphasis=0.0,
    ref_level_db=20,
    log_func="np.log",
    do_trim_silence=False,
    trim_db=45,
    mel_fmin=0,
    mel_fmax=None,
    spec_gain=1.0,
    signal_norm=False,
    do_amp_to_db_linear=False,
)
config = VitsConfig(
    audio=audio_config,
    run_name="vits_vctk_en_us",
    batch_size=48,
    eval_batch_size=16,
    batch_group_size=5,
    num_loader_workers=4,
    num_eval_loader_workers=4,
    run_eval=True,
    test_delay_epochs=-1,
    epochs=10000,
    text_cleaner="english_cleaners",
    use_phonemes=True,
    phoneme_language="en-us",
    phoneme_cache_path=os.path.join(output_path, "phoneme_cache"),
    compute_input_seq_cache=True,
    print_step=25,
    print_eval=True,
    mixed_precision=True,
    max_seq_len=500000,
    output_path=output_path,
    datasets=[dataset_config],
)
args, config, output_path, _, c_logger, tb_logger = init_training(TrainingArgs(), config)
trainer = Trainer(args, config, output_path, c_logger, tb_logger, cudnn_benchmark=True)
trainer.fit()`

**Reply by erogol** [reply]
@skol101 I feel your pain right now as I try multi-speaker training myself :)). Our multi-speaker API is a 💩. 

I am doing a bit of refactoring hoping to make things better. 

I am also creating a new VCTK recipe for VITS. Once it is done you can use it for your run. Sorry for it :(

**Reply by skol101** [reply]
Great! Looking forward.

@erogol was vits_vctk model (that's now available for downloading) trained using multi-speaker API?

**Reply by erogol** [reply]
yes it was.

**Reply by MahdiEsrafili** [reply]
> yeah right point. The model is with only the generator.


How is it available to save generator part of my own VITS trained model?
